### PR TITLE
DSDEEPB-852 Provide web endpoint to abort a workflow.

### PR DESCRIPTION
**Pull request by mcovarr**

### CODE REVIEW

File: src/main/scala/cromwell/engine/package.scala [hunk 1]
```diff
@@ -19,7 +19,7 @@ package object engine {
   type WorkflowId = UUID
 
   sealed trait WorkflowState {
-    def isTerminal: Boolean
+    val isTerminal: Boolean
```
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/package.scala]: Having the trait define things as `def` and the implementors going with `val`s is a good thing IMO as it allows for an implementor to make it whatever they want

Comment (mcovarr) [inline src/main/scala/cromwell/engine/package.scala]: makes sense, will fix

Comment (geoffjentry) [inline src/main/scala/cromwell/engine/package.scala]: I was curious if that was intentional despite what I said, it sounds like "no" :)

I think there *are* cases where a trait should enforce val-ness and I was on the fence on if I thought this is one of them. I sided with "no" but could be easily swayed to "yes"

Comment (mcovarr) [inline src/main/scala/cromwell/engine/package.scala]: I was trying to make it consistent, but your argument for why it was right the way it was before makes more sense than consistency.  :smile:

**Comment by mcovarr** [comment]
Er whoops totally forgot to update the docs

File: src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala [hunk 2]
```diff
@@ -101,8 +109,46 @@ class CromwellApiServiceSpec extends FlatSpec with CromwellApiService with Scala
...
+    Post(s"/workflow/foobar/abort") ~>
+      abortRoute ~>
+      check {
+        assertResult(StatusCodes.BadRequest) {
+          status
+        }
+      }
+  }
+
+  it should "return 200 for abort of a known workflow id" in {
```
Comment (cjllanwarne) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: 200 when the abort succeeds. Should there be an alternative response code if the abort fails or ended with the horrible ''might have worked' state?

File: src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala [hunk 3]
```diff
@@ -101,8 +109,46 @@ class CromwellApiServiceSpec extends FlatSpec with CromwellApiService with Scala
...
+      abortRoute ~>
+      check {
+        assertResult(StatusCodes.OK) {
+          status
+        }
+
+        assertResult(
+          s"""{
+             |  "id": "${MockWorkflowManagerActor.runningWorkflowId.toString}",
+             |  "status": "Aborted"
```
Comment (cjllanwarne) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: Does this status field have a value in all of the above situations or is everything implied via response code?

Comment (mcovarr) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: Currently everything is implied via response code.

File: src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala (hunk 2 above)
Comment (mcovarr) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: Perhaps, depending on how our abort logic ends up being written.  Currently this is very simplistic and just validates that the supplied ID looks like a workflow ID and corresponds to a known "running" workflow.  There's currently no test for what the right behavior is when the ID corresponds to a known workflow in a terminal state.

**Comment by cjllanwarne** [comment]
OK by me, I'll throw the second review responsibility to @kshakir

File: src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala (hunk 2 above)
Comment (mcovarr) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: I'd like to update this to handle terminal workflows, does 403 seem like the right status?

Comment (geoffjentry) [inline src/test/scala/cromwell/webservice/CromwellApiServiceSpec.scala]: Perhaps 405, although by the rest bible it's not completely right. However the actual description is spot on.

**Comment by kshakir** [comment]
Post rebase, please:
- change `/workflow` to `/workflows`
- add the `version` to the making it `/workflows/:version/:id/abort`
- add the response annotation to the `200` response
- double check that you can create and send an abort request via the ui (`sbt 'run server'`, then visit [/swagger](http://localhost:8000/swagger))

With the edits above this PR looks good to go for me. Besides that, the un-DRYness of swagger and akka continue to make my skin crawl, but I don't know any better best practices.

**Comment by mcovarr** [comment]
OK to all suggestions except the 405 for attempted aborts on terminal workflows.  I've only seen that used for inappropriate HTTP verbs, which Spray should already be checking before it gets to our business logic.

FYI there's some interesting work being done by the Rawls team to get the Swagger noise out of the code, although it seems to remain as un-DRY as ever:

https://github.com/broadinstitute/rawls/pull/63/files

**Comment by geoffjentry** [comment]
I have a better idea on how to get the swagger noise out of the code .... :)

**Comment by mcovarr** [comment]
Updated per @kshakir's AC above.  :smile: 

The "produces" content type is not 100% right in the Swagger annotations here and probably in other APIs.  This API is documented as producing application/json which it will for successful requests, but for failed requests text/plain is produced instead.  Given the way the "Try it out!"s are built, they will fail with undocumented 406s if the underlying request fails with a documented status.

**Pull request merged**